Objects/object.c

```diff
@@ -2545,6 +2528,25 @@ _Py_ForgetReference(PyObject *op)
 
     PyInterpreterState *interp = _PyInterpreterState_GET();
 
+    if (!_PyRefchain_IsTraced(interp, op)) {
+        if (PyUnicode_CHECK_INTERNED(op)) {
+            // interned strings can be shared between the main interpreter and
+            // between sub-interpreters due to the shared interp dict.  See
+            // init_interned_dict().  In this case, the string was created and
+            // traced by a different sub-interpreter.
+            return;
+        }
+        PyInterpreterState *main_interp = _PyInterpreterState_Main();
+        if (interp != main_interp  &&
+            interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC) {
```

Comment:
FYI, you can also use `_Py_IsMainInterpreter()`:

```suggestion
        if (_Py_IsMainInterpreter(interp)  &&
            interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC) {
```